include/gz/math/TimeVaryingVolumetricGridLookupField.hh

```diff
@@ -0,0 +1,291 @@
+        const S &_session,
+        const std::vector<InterpolationPoint4D<T, V>> &_points,
+        const std::vector<X> &_values1,
+        const std::vector<X> &_values2,
+        const X _default = X(0)
+      ) const;
+    };
+
+    /// \brief An in-memory session. Loads the whole dataset in memory and
+    /// performs queries.
+    template<typename T, typename V>
+    class InMemorySession
+    {
+      /// \brief Iterator which holds pointer to current state
+      /// TODO(arjo): Use friend to make visible only to InMemorySession
+      /// specialization
+      private:
+        typename std::map<T, VolumetricGridLookupField<V>>::const_iterator iter;
+
+      /// \brief Time of last query
+      public: T time;
+
+      friend class
+        TimeVaryingVolumetricGridLookupField<T, V, InMemorySession<T, V>>;
+    };
+
+    /// \brief Specialized version of `TimeVaryingVolumetricGridLookupField`
+    /// for in-memory lookup. It loads the whole dataset into memory.
+    template<typename T, typename V>
+    class TimeVaryingVolumetricGridLookupField<T, V, InMemorySession<T, V>>
+    {
+      /// \brief Default constructor
+      public: TimeVaryingVolumetricGridLookupField()
+      {}
+
+      /// \brief Documentation inherited
+      public: void AddVolumetricGridField(
+        const T &_time, const VolumetricGridLookupField<V> &_field) {
+        this->gridFields.emplace(_time, _field);
+      }
+
+      /// \brief Documentation inherited
+      public: InMemorySession<T, V> CreateSession() const {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.begin();
+        sess.time = T(0);
+        return sess;
+      }
+
+      /// \brief Documentation inherited
+      public: InMemorySession<T, V> CreateSession(const T &_time) const {
+        InMemorySession<T, V> sess;
+        sess.iter = this->gridFields.lower_bound(_time);
+        sess.time = _time;
+        return sess;
+      }
+
+      /// \brief Documentation inherited
+      public: std::optional<InMemorySession<T, V>> StepTo(
+        const InMemorySession<T, V> &_session, const T &_time) const {
+        if (_session.iter == gridFields.end())
+        {
+          return std::nullopt;
+        }
+
+        InMemorySession<T, V> newSess(_session);
+
+        auto nextTime = std::next(_session.iter);
+        if (nextTime == this->gridFields.end() || _time < _session.iter->first)
+        {
+          return std::nullopt;
+        }
+
+        while (nextTime != this->gridFields.end()
+          && nextTime->first <= _time)
+        {
+          newSess.iter = nextTime;
+          nextTime = std::next(nextTime);
+        }
+        newSess.time = _time;
+        return newSess;
+      }
+
+      /// \brief Documentation inherited
+      public: std::vector<InterpolationPoint4D<T, V>>
+        LookUp(const InMemorySession<T, V> &_session,
+          const Vector3<V> &_point,
+          const Vector3<V> &_tol = Vector3<V>{0.5, 0.5, 0.5}) const {
```

Comment:
@arjo129 meta: depending on the scale and nature of `V` values, `0.5` can be huge for a tolerance. I don't mind strongly about the arbitrary value, but I do think there should be a way to change this tolerance when using a `TimeVaryingVolumetricGrid`.